File: src/main/java/cn/huangsy/DailyTemperatures.java
package cn.huangsy;

/**
 * @author huangsy
 * @date 2020/6/11 9:10
 */
public class DailyTemperatures {

    public int[] dailyTemperatures(int[] T) {
        int[] result = new int[T.length];
        int[] dp = new int[101];
        for (int i = T.length - 2; i >= 0; i--) {
            int right = i + 1;
            if (T[right] > T[i]) {
                result[i] = 1;
            } else {
                do {
                    result[i] += dp[T[right]];
                    right += dp[T[right]];
                } while (right < T.length && T[right] <= T[i] && dp[T[right]]!=0);
                if (result[i] != 0) {
                    result[i] += 1;
                }
                if (dp[T[right]]==0 && T[right] <= T[i]) {
                    result[i] = 0;
                }
            }
            dp[T[i]] = result[i];
        }
        return result;
    }

    public static void main(String[] args) {
        DailyTemperatures dailyTemperatures = new DailyTemperatures();
        System.out.println(dailyTemperatures.dailyTemperatures(new int[]{73, 74, 75, 71, 69, 72, 76, 73}));
        System.out.println(dailyTemperatures.dailyTemperatures(new int[]{55,38,53,81,61,93,97,32,43,78}));
    }
}

File: src/main/java/cn/leetcode/739/DailyTemperatures.kt
package cn.leetcode.`739`

/**
 * @author huangsy
 * @date 2020/6/11 10:32
 */
class DailyTemperatures {

    fun dailyTemperatures(T: IntArray): IntArray {
        val result = IntArray(T.size)
        for (i in T.size - 2 downTo 0) {
            var right = i + 1
            if (T[right] > T[i]) {
                result[i] = 1
            } else {
                while (true) {
                    if (T[right] > T[i]) {
                        result[i] = right - i;
                        break;
                    } else if (result[right] == 0) {
                        result[i] = 0;
                        break;
                    }
                    right += result[right];
                }
            }
        }
        return result
    }
}